### PR TITLE
fix(invity): Spend - transaction not shown in Trade tab history

### LINKED ISSUE #3276: Spend - transaction not shown in Trade tab history
**Describe the bug**
Outgoing transaction is not displayed in the shared Trade tab transaction history, only as a generic send transaction in the Overview tab.

**Info:**
 - Suite Version: web 21.1.0 develop, commit fa7d200c
 - Browser: Chromium 87.0.4280.66
 - OS: Nixos 20.09
 - Firmware Version: model T 2.3.4 50854b921

**Steps to reproduce:**
1. Navigate to Accounts -> Trade -> Spend,
2. Complete the spend flow successfully,
3. Observe.

**Expected result:**
The outgoing transaction is displayed in the Trade tab transaction history.

**Actual result:**
The outgoing transaction is displayed in the Overview tab without any indication it is the one that was used in the Spend flow.

**Screenshots:**
image
image

### PR DESCRIPTION
**Pull request by LukasRada**
fixes #3276 + refactor saving and displaying buy/exchange/crypto trades

### CODE REVIEW

File: packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/List/Quote/index.tsx [hunk 1]
```diff
@@ -3,10 +3,10 @@ import styled from 'styled-components';
 import { Button, variables, Icon, useTheme } from '@trezor/components';
 import { QuestionTooltip, Translation } from '@suite-components';
 import { ExchangeTrade } from 'invity-api';
-import CoinmarketExchangeProviderInfo from '@wallet-components/CoinmarketExchangeProviderInfo';
 import { formatCryptoAmount } from '@wallet-utils/coinmarket/coinmarketUtils';
 import { isQuoteError } from '@wallet-utils/coinmarket/exchangeUtils';
 import { useCoinmarketExchangeOffersContext } from '@wallet-hooks/useCoinmarketExchangeOffers';
+import CoinmarketProviderInfo from '@suite/components/wallet/CoinmarketProviderInfo';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/List/Quote/index.tsx]:
```suggestion
import { CoinmarketProviderInfo } from '@wallet-components';
```

File: packages/suite/src/storage/definitions.ts [hunk 2]
```diff
@@ -8,7 +8,8 @@ import { AcquiredDevice } from '@suite-types';
 import { MetadataState } from '@suite-types/metadata';
 import { Account, Discovery, CoinFiatRates, WalletAccountTransaction } from '@wallet-types';
 import { GraphData } from '@wallet-types/graph';
-import { BuyTrade, ExchangeTrade } from 'invity-api';
+import type { TradeType } from '@wallet-types/coinmarketCommonTypes';
```
Comment (szymonlesisz) [inline packages/suite/src/storage/definitions.ts]:
```suggestion
import { TradeType } from '@wallet-types/coinmarketCommonTypes';
```

File: packages/suite/src/actions/suite/storageActions.ts [hunk 3]
```diff
@@ -10,13 +10,16 @@ import * as suiteActions from '@suite-actions/suiteActions';
 import { serializeDiscovery, serializeDevice } from '@suite-utils/storage';
 import { deviceGraphDataFilterFn } from '@wallet-utils/graphUtils';
 import { FormState } from '@wallet-types/sendForm';
-import { BuyTrade, ExchangeTrade } from 'invity-api';
+import type { TradeType } from '@wallet-types/coinmarketCommonTypes';
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/storageActions.ts]:
```suggestion
import { TradeType } from '@wallet-types/coinmarketCommonTypes';
```

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
+import { TradeSpend } from '@suite/reducers/wallet/coinmarketReducer';
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx]:
```suggestion
import { TradeSpend } from '@wallet-reducers/coinmarketReducer';
```
and pls move it below

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import React from 'react';
+import styled from 'styled-components';
+import { Translation, HiddenPlaceholder, FormattedDate } from '@suite-components';
+import { variables } from '@trezor/components';
+import Status from '../Status';
+import { CoinmarketProviderInfo } from '@suite/components/wallet';
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx]:
```suggestion
import { CoinmarketProviderInfo } from '@wallet-components';
```

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts [hunk 6]
```diff
@@ -157,30 +157,18 @@ const storageMiddleware = (api: MiddlewareAPI<Dispatch, AppState>) => (next: Dis
...
-                action.date,
-            );
-            break;
-
-        case COINMARKET_EXCHANGE.SAVE_TRADE:
-            storageActions.saveExchangeTrade(
-                action.data,
+        case COINMARKET_COMMON.SAVE_TRADE:
+            storageActions.saveCoinmarketTrade(
+                action.data as Trade,
```
Comment (szymonlesisz) [inline packages/suite/src/middlewares/wallet/storageMiddleware.ts]: is type casting necessary?

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts [hunk 7]
```diff
@@ -19,6 +18,7 @@ import { WalletAction } from '@wallet-types';
 import { getDiscovery } from '@wallet-actions/discoveryActions';
 import { isDeviceRemembered } from '@suite-utils/device';
 import { serializeDiscovery } from '@suite-utils/storage';
+import { Trade } from '@suite-actions/storageActions';
```
Comment (szymonlesisz) [inline packages/suite/src/middlewares/wallet/storageMiddleware.ts]: ambiguous names/types `Trade` in storageActions and coinmarketReducer.
maybe both should be placed in shareable file `@wallet-types/coinmarketCommonTypes` ?

File: packages/suite/src/reducers/wallet/coinmarketReducer.ts [hunk 8]
```diff
@@ -35,7 +36,8 @@ type CommonTrade = {
 
 export type TradeBuy = CommonTrade & { tradeType: 'buy'; data: BuyTrade };
 export type TradeExchange = CommonTrade & { tradeType: 'exchange'; data: ExchangeTrade };
-export type Trade = TradeExchange | TradeBuy;
+export type TradeSpend = CommonTrade & { tradeType: 'spend'; data: SpendTrade };
+export type Trade = TradeExchange | TradeBuy | TradeSpend;
```
Comment (szymonlesisz) [inline packages/suite/src/reducers/wallet/coinmarketReducer.ts]: commented above, consider move all types to common file

**Review by szymonlesisz — CHANGES_REQUESTED** [review]
looks great, didn't test it however

just few nitpicks regarding imports
you probably want to disable auto import plugin for this project :)

File: packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/List/Quote/index.tsx (hunk 1 above)
Comment (LukasRada) [reply packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/List/Quote/index.tsx]: done

File: packages/suite/src/storage/definitions.ts (hunk 2 above)
Comment (LukasRada) [reply packages/suite/src/storage/definitions.ts]: done

File: packages/suite/src/actions/suite/storageActions.ts (hunk 3 above)
Comment (LukasRada) [reply packages/suite/src/actions/suite/storageActions.ts]: done

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx (hunk 5 above)
Comment (LukasRada) [reply packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx]: done

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts (hunk 6 above)
Comment (LukasRada) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: Now it is not :) done

File: packages/suite/src/middlewares/wallet/storageMiddleware.ts (hunk 7 above)
Comment (LukasRada) [reply packages/suite/src/middlewares/wallet/storageMiddleware.ts]: done

File: packages/suite/src/reducers/wallet/coinmarketReducer.ts (hunk 8 above)
Comment (LukasRada) [reply packages/suite/src/reducers/wallet/coinmarketReducer.ts]: done

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx (hunk 4 above)
Comment (LukasRada) [reply packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SpendTransaction/index.tsx]: done

**Review by szymonlesisz — APPROVED** [review]
great, you can squash it now

**Comment by bosomt** [comment]
**QA OK**

https://github.com/trezor/trezor-suite/issues/3276#issuecomment-813246865

**Pull request merged**